packages/shadydom/src/patches/ParentNode.js

```diff
@@ -128,24 +131,248 @@ export const ParentNodePatches = utils.getOwnPropertyDescriptors({
+   * eventually results in a complete match.
+   *
+   * - `complexSelectorParts` is the decomposed version of the selector that
+   * this cursor is attempting to match.
+   *
+   * - `position` is an element that matches a compound selector in
+   * `complexSelectorParts`.
+   *
+   * - `index` is the index of the compound selector in `complexSelectorParts`
+   * that matched `position`.
+   *
+   * @typedef {{
+   *   target: !Element,
+   *   complexSelectorParts: !ComplexSelectorParts,
+   *   position: !Element,
+   *   index: number,
+   * }}
+   */
+  let SelectorMatchingCursor;
+
+  /**
+   * The list of `SelectorMatchingCursor`s, initialized with cursors pointing at
+   * all descendants of `contextElement` that match the last compound selector
+   * in any complex selector in `selectorList`.
+   *
+   * @type {!Array<!SelectorMatchingCursor>}
+   */
+  let cursors = utils.flat(
+    query(contextElement, (_element) => true).map((element) => {
+      return utils.flat(
+        complexSelectors.map((complexSelectorParts) => {
+          const {compoundSelectors} = complexSelectorParts;
+          const index = compoundSelectors.length - 1;
+          if (matchesCompoundSelector(element, compoundSelectors[index])) {
+            return [
+              {
+                target: element,
+                complexSelectorParts,
+                position: element,
+                index,
+              },
+            ];
+          } else {
+            return [];
+          }
+        })
+      );
+    })
+  );
+
+  // At each step, any remaining cursors that have not finished matching (i.e.
+  // with `cursor.index > 0`) should be replaced with new cursors for any valid
+  // candidates that match the next compound selector.
+  while (cursors.length > 0 && cursors.some((cursor) => cursor.index > 0)) {
+    cursors = utils.flat(
+      cursors.map((cursor) => {
+        // Cursors with `index` of 0 have already matched and should not be
+        // replaced or removed.
+        if (cursor.index <= 0) {
+          return cursor;
+        }
+
+        const {
+          target,
+          position,
+          complexSelectorParts,
+          index: lastIndex,
+        } = cursor;
+        const index = lastIndex - 1;
+        const combinator = complexSelectorParts.combinators[index];
+        const compoundSelector = complexSelectorParts.compoundSelectors[index];
+
+        if (combinator === ' ') {
+          const results = [];
+
+          // For `a b`, where existing cursors have `position`s matching `b`,
+          // the candidates to test against `a` are all ancestors each cursor's
+          // `position`.
+          for (
+            let ancestor = position[utils.SHADY_PREFIX + 'parentNode'];
+            ancestor && ancestor instanceof Element;
+            ancestor = ancestor[utils.SHADY_PREFIX + 'parentNode']
+          ) {
+            if (matchesCompoundSelector(ancestor, compoundSelector)) {
```

Comment:
It wasn't obvious to me why results would ever be an array so maybe this should be commented, but the reason is that there may be > 1 candidate in the ancestor tree...

e.g.: given `c c c d` and `<c><c><c><d>`: the first cursor will be for `d` and next we need a cursor for *each* `c`, but eventually the one for the top most `c` will survive. Is that right?

Reply:
Yeah, each ancestor matching `c` needs a cursor because any of them could potentially be part of the match when it's only gotten around to testing the rightmost `c` in the selector. The cursors that point at the shallower `<c>`s will eventually be unable to find enough candidate `<c>` ancestors to match the rest of the selector and will be replaced with zero cursors.

That particular case will look something like this:
```
<c>
  <c>
    <c>
      <d> <!-- cursor[0] `c c c (d)`-->
```
```
<c> <!-- candidate for: cursor[0] `c c c( )d` -->
  <c> <!-- candidate for: cursor[0] `c c c( )d` -->
    <c> <!-- candidate for: cursor[0] `c c c( )d` -->
      <d>
```
```
<c> <!-- cursor[2] `c c (c) d` -->
  <c> <!-- cursor[1] `c c (c) d` -->
    <c> <!-- cursor[0] `c c (c) d` -->
      <d>
```
During this step, `cursor[2]` doesn't find any candidates:
```
<c> <!-- candidate for: cursor[0] `c c( )c d`, cursor[1] `c c( )c d` -->
  <c> <!-- candidate for: cursor[0] `c c( )c d` -->
    <c>
      <d>
```
```
<c> <!-- cursor[1] `c (c) c d`, cursor[2] `c (c) c d` -->
  <c> <!-- cursor[0] `c (c) c d` -->
    <c>
      <d>
```
During this step, `cursor[1]` doesn't find any candidates:
```
<c> <!-- candidate for: cursor[0] `c( )c c d` -->
  <c>
    <c>
      <d>
```
```
<c> <!-- cursor[0]: `(c) c c d` -->
  <c>
    <c>
      <d>
```

Reply:
> The cursors that point at the shallower `<c>`s will eventually be unable to find enough candidate `<c>` ancestors to match the rest of the selector and will be replaced with zero cursors.

Or, maybe I should have said "Cursors with 'algorithmic ancestors' that pointed at the shallower `<c>`s will be unable..." since only cursors that have already succeeded in making a complete match match actually 'survive' between iterations - all others are replaced with zero or more new cursors. The overlapping tree / lineage terminology that's shared here between the DOM and the timeline of cursors makes describing this a bit difficult. :/